### PR TITLE
Let `Range#step` delegate to `begin.step`

### LINKED ISSUE #9339: Float-based ranges
Proposal: It would be help in many situations to make the Range(Float64, Float64) and Range(Float32, Float32) types enumerable with a step size (default to 1). 

One potential pain point is the constructor requiring three arguments instead of 2 now so the X..Y syntax won't work if one wants to specify the delta.

Looking at the code in "range.cr" I think a simple way to get at the enumeration is to put an internal function to the range that does a responds_to?(:succ) on the current range element to either use the provided method or to try and infer it by doing something like current + step_size. A slight downside of this is that now the underlying type needs an operator+ defined.

### PR DESCRIPTION
**Pull request by straight-shoota**
TLDR; This lets `(1.0..5.0).step` work.

Resolves #9339
Precursers: #10203, #10130

`Range#step` previously was based on calling `#succ` to get the next value in the iteration. This works for discrete types like integers which define `#succ`. Floating point numbers don't have a clear successor, so `Float#succ` doesn't exist and `Range(Float64, Float64)#step` wouldn't work.

This patch lets `Range(B, E)#step` delegate to `B#step` (if defined), so the behaviour can be entirely implemented by the value's type. The expected interface is that of `Number#step` which is also extended in this patch by an `exclusive` argument to handle exclusive ranges.

The implementation of `Number#step` is very generic based on simple algebra methods and works for non-number types as well. So a follow up would extract it for re-usability (an example would be `Time` as per #9327).

Note: Non-discrete types don't have an inherent step size. For stepping a `Time` range for example, you would need to specifiy the step size explicitly. `Range#step` uses `1` as default value for the step size, but if the step implementation doesn't accept integer step size, it errors at compile time. The error message doesn't point exactly to the call to `Range#step`, but I guess that's acceptable.

The previous `#succ`-based implementation is retained as an alternative if `begin` doesn't respond to `#step`. This is still used for `String` ranges. I'm not sure if we want to keep this alternative. It's not really necessary, `Range` can just expect every stepable type to implement `#step`. But this is a more complex discussion, because `#succ` and `#pred` methods are currently the defining interface of value types for many methods on `Range`.

So, for now I would just add this as an enhancement. The diff in `range.cr` is actually pretty minimal, it's best viewed with whitespace option.

### CODE REVIEW

File: src/number.cr [hunk 1]
```diff
@@ -239,6 +239,29 @@ struct Number
         @current += @step
       end
     end
+
+    # Overrides `Enumerable#sum` to use more performant implementation on integer
+    # ranges.
+    def sum(initial)
```
Comment (asterite) [inline src/number.cr:245]: Is there an existing spec for this?

File: src/number.cr [hunk 2]
```diff
@@ -239,6 +239,29 @@ struct Number
...
+      super if @reached_end
```
Comment (asterite) [inline src/number.cr:246]: Shouldn't this return?

File: src/number.cr (hunk 1 above)
Comment (straight-shoota) [reply src/number.cr:245]: Yes, I just copied this method from `Range::StepIterator` because the sum specs in `range_spec` happen to call this method. Especially these specs runs *very* long without it:

https://github.com/crystal-lang/crystal/blob/cb09b0e218768c759f261424f10a0abb48edff0f/spec/std/range_spec.cr#L110-L111

File: src/number.cr (hunk 2 above)
Comment (straight-shoota) [reply src/number.cr:246]: Haha, yeah. That's just copied 1:1 from `Range::StepIterator` 😆

**Review by asterite — APPROVED** [review]

**Comment by asterite** [comment]
Can't wait to step through Time!

But what's the default step there?

**Comment by straight-shoota** [comment]
None. I mentioned this in the OP. Stepping through non-discrete types like time requires a `step` argument.
